File: clojure/3.clj
(load-file "lib/core.clj")
(use
  '[clojure.set :only (union)]
  '[lib.core :only (ceil round sqrt)])


(defn fermat-factors
  [n]
  (if (= 0 (rem n 2))
    (sorted-set n (/ n 2))
    (loop [a (ceil (sqrt n))]
      (let [b (- (* a a) n)
            s (sqrt b)]
        (if (== s (round s))
          (sorted-set (- a s) (+ a s))
          (recur (inc a)))))))

(defn prime-factors
  [n]
  (let [factors (fermat-factors n)]
    (if (contains? factors 1.0)
      (disj factors 1.0)
      (reduce union (map prime-factors factors)))))

(def result (long (reduce max (prime-factors 600851475143))))

(println result)
(assert (= result 6857))

File: solns/ruby/3.rb
def fermat_factors(n)
  return [2, n/2] if n % 2 == 0
  a = Math.sqrt(n).ceil
  b = a*a-n
  s = Math.sqrt(b)
  while s != s.round
    a += 1
    b = a*a-n
    s = Math.sqrt(b)
  end
  [a-s, a+s]
end

def prime_factors(n)
  factors = fermat_factors(n)
  if factors.include? 1
    factors.delete_if { |f| f == 1 }
  else
    factors.map { |f| prime_factors(f) }.inject(&:concat)
  end
end

result = prime_factors(600851475143).max.round

puts result
raise Error unless result == 6857
